src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,524 @@ extern "C" {
+		return false;
+	}
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GetGroupUniqueTleReferencesWalker, context);
+};
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingFuncRewriteWalker
+//
+//	@doc:
+//		Goes over the expression tree and update arguments of GROUPING
+//		function.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingFuncRewriteWalker(
+	Node *node, SContexGroupingFuncRewriteWalker *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, GroupingFunc))
+	{
+		GroupingFunc *gf = (GroupingFunc *) node;
+		ListCell *arg_lc;
+		List *newargs = NIL;
+
+		gf->ngrpcols = context->m_ngrpcols;
+
+		// For each argument in gf->args, find its new position
+		ForEach(arg_lc, gf->args)
+		{
+			int arg = intVal(lfirst(arg_lc));
+			int new_arg =
+				gpdb::ListNthInt(context->m_grouping_tle_refs_mapping, arg);
+			newargs = gpdb::LAppend(newargs, gpdb::MakeIntegerValue(new_arg));
+		}
+
+		// Free gf->args since we do not need it any more.
+		gpdb::ListFreeDeep(gf->args);
+		gf->args = newargs;
+		return false;
+	}
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GroupingFuncRewriteWalker, context);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GetVarsWithoutTleWalker
+//
+//	@doc:
+//		Goes over the expression tree, finds all vars that do not have relevant
+//		targetlist entries, and adds resjunk targetlist entries for such vars
+//		to a m_tlist_addition in the context, that later will be added to the
+//		original target list.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GetVarsWithoutTleWalker(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, Aggref))
+	{
+		// do not examine what is inside Aggref
+		return false;
+	}
+
+	// find all vars that do not have relevant targetlist entries, and add
+	// resjunk targetlist entries for such vars to m_tlist_addition
+	if (IsA(node, Var))
+	{
+		bool found = false;
+		Var *var = (Var *) gpdb::CopyObject(node);
+		if (var->varlevelsup == context->m_current_query_level)
+		{
+			var->varlevelsup = 0;
+			ListCell *lc_tle = NULL;
+			ForEach(lc_tle, context->m_query->targetList)
+			{
+				TargetEntry *target_entry = (TargetEntry *) lfirst(lc_tle);
+
+				if (IsA(target_entry->expr, Var))
+				{
+					if (gpdb::Equals(target_entry->expr, var))
+					{
+						found = true;
+						break;
+					}
+				}
+			}
+			if (!found)
+			{
+				int ressno = gpdb::ListLength(context->m_query->targetList) + 1;
+				TargetEntry *newTargetEntry =
+					gpdb::MakeTargetEntry((Expr *) var, ressno, NULL, true);
+
+				context->m_tlist_addition =
+					gpdb::LAppend(context->m_tlist_addition, newTargetEntry);
```

Comment:
Why not add directly to context->m_query->targetList, avoiding the context->m_tlist_addition field and list concatenation at the end?

Reply:
Adding directly to context->m_query->targetList would contradict the idea of expression_tree_walker, please refer to its documentation:
```
 * expression_tree_walker() is designed to support routines that traverse
 * a tree in a read-only fashion (although it will also work for routines
 * that modify nodes in-place but never add/delete/replace nodes)
```